// @storekeeper/escpos/lib/src/escpos/escposCommands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMAND_MAP = exports.EscPos = void 0;
var textCommand_1 = require("./commands/textCommand");
var tableCommand_1 = require("./commands/tableCommand");
var barcodeCommand_1 = require("./commands/barcodeCommand");
var lineCommand_1 = require("./commands/lineCommand");
var lineFeedCommand_1 = require("./commands/lineFeedCommand");
var rawCommand_1 = require("./commands/rawCommand");
var cutCommand_1 = require("./commands/cutCommand");
var imageCommand_1 = require("./commands/imageCommand");
var lineSpacingCommand_1 = require("./commands/lineSpacingCommand");
var qrCommand_1 = require("./commands/qrCommand");
exports.EscPos = {
    // Printer hardware
    HW_INIT: [0x1b, 0x40],
    HW_SELECT: [0x1b, 0x3d, 0x01],
    HW_RESET: [0x1b, 0x3f, 0x0a, 0x00],
    // Feed control sequences
    CTL_LF: [0x0a],
    CTL_FF: [0x0c],
    CTL_CR: [0x0d],
    CTL_HT: [0x09],
    CTL_VT: [0x0b],
    // Paper
    PAPER_FULL_CUT: [0x1d, 0x56, 0x00],
    PAPER_PARTIAL_CUT: [0x1d, 0x56, 0x01],
    PAPER_CUT_A: [0x1d, 0x56, 0x41],
    PAPER_CUT_B: [0x1d, 0x56, 0x42],
    // Cash Drawer
    CD_KICK_2: [0x1b, 0x70, 0x00],
    CD_KICK_5: [0x1b, 0x70, 0x01],
    // Code Pages
    CP_SET: [0x1b, 0x74],
    CP_CP437: [0x0],
    CP_CP850: [0x2],
    CP_CP860: [0x3],
    CP_CP863: [0x4],
    CP_CP865: [0x5],
    CP_CP1251: [0x6],
    CP_CP866: [0x7],
    CP_MACCYRILLIC: [0x8],
    CP_CP775: [0x9],
    CP_CP1253: [0x10],
    CP_CP737: [0x11],
    CP_CP857: [0x12],
    CP_ISO8859_9: [0x13],
    CP_CP864: [0x14],
    CP_CP862: [0x15],
    CP_ISO8859_2: [0x16],
    CP_CP1253_ALT: [0x17],
    CP_CP1250: [0x18],
    CP_CP858: [0x19],
    CP_CP1254: [0x20],
    CP_CP737_ALT: [0x24],
    CP_CP1257: [0x25],
    CP_CP847: [0x26],
    CP_CP885: [0x28],
    CP_CP857_ALT: [0x29],
    CP_CP1250_ALT: [0x30],
    CP_CP775_ALT: [0x31],
    CP_CP1254_ALT: [0x32],
    CP_CP1256: [0x34],
    CP_CP1258: [0x35],
    CP_ISO8859_2_ALT: [0x36],
    CP_ISO8859_3: [0x37],
    CP_ISO8859_4: [0x38],
    CP_ISO8859_5: [0x39],
    CP_ISO8859_6: [0x40],
    CP_ISO8859_7: [0x41],
    CP_ISO8859_8: [0x42],
    CP_ISO8859_9_ALT: [0x43],
    CP_ISO8859_15: [0x44],
    CP_CP856: [0x47],
    CP_CP874: [0x47],
    // Text formating
    TXT_SIZE: [0x1d, 0x21],
    TXT_NORMAL: [0x1b, 0x21, 0x00],
    TXT_2HEIGHT: [0x1b, 0x21, 0x10],
    TXT_2WIDTH: [0x1b, 0x21, 0x20],
    TXT_4SQUARE: [0x1b, 0x21, 0x30],
    TXT_UNDERL_OFF: [0x1b, 0x2d, 0x00],
    TXT_UNDERL_ON: [0x1b, 0x2d, 0x01],
    TXT_UNDERL2_ON: [0x1b, 0x2d, 0x02],
    TXT_BOLD_OFF: [0x1b, 0x45, 0x00],
    TXT_BOLD_ON: [0x1b, 0x45, 0x01],
    TXT_FONT_A: [0x1b, 0x4d, 0x00],
    TXT_FONT_B: [0x1b, 0x4d, 0x01],
    TXT_ALIGN_LT: [0x1b, 0x61, 0x00],
    TXT_ALIGN_CT: [0x1b, 0x61, 0x01],
    TXT_ALIGN_RT: [0x1b, 0x61, 0x02],
    TXT_INVERT_ON: [0x1d, 0x42, 0x01],
    TXT_INVERT_OFF: [0x1d, 0x42, 0x00],
    TXT_COLOR_BLACK: [0x1b, 0x72, 0x00],
    TXT_COLOR_RED: [0x1b, 0x72, 0x01],
    // Barcodes
    BARCODE_TXT_OFF: [0x1d, 0x48, 0x00],
    BARCODE_TXT_ABV: [0x1d, 0x48, 0x01],
    BARCODE_TXT_BLW: [0x1d, 0x48, 0x02],
    BARCODE_TXT_BTH: [0x1d, 0x48, 0x03],
    BARCODE_FONT_A: [0x1d, 0x66, 0x00],
    BARCODE_FONT_B: [0x1d, 0x66, 0x01],
    BARCODE_HEIGHT: [0x1d, 0x68],
    BARCODE_WIDTH: [0x1d, 0x77],
    BARCODE_UPC_A: [0x1d, 0x6b, 0x00],
    BARCODE_UPC_E: [0x1d, 0x6b, 0x01],
    BARCODE_EAN13: [0x1d, 0x6b, 0x02],
    BARCODE_EAN8: [0x1d, 0x6b, 0x03],
    BARCODE_CODE39: [0x1d, 0x6b, 0x04],
    BARCODE_ITF: [0x1d, 0x6b, 0x05],
    BARCODE_NW7: [0x1d, 0x6b, 0x06],
    BARCODE_CODE128: [0x1d, 0x6b, 0x49],
    LEFT_MARGIN: [0x1d, 0x4c],
    PRINT_AREA_WIDTH: [0x1d, 0x57],
    LINE_CHAR: [0xc4],
    BITMAP_D24: [0x1b, 0x2a, 0x21],
    LS_DEFAULT: [0x1b, 0x32],
    LS_SET: [0x1b, 0x33]
};
exports.COMMAND_MAP = {
    'text': textCommand_1.TextCommand,
    'table': tableCommand_1.TableCommand,
    'barcode': barcodeCommand_1.BarcodeCommand,
    'line': lineCommand_1.LineCommand,
    'linefeed': lineFeedCommand_1.LineFeedCommand,
    'lineSpacing': lineSpacingCommand_1.LineSpacingCommand,
    'raw': rawCommand_1.RawCommand,
    'cut': cutCommand_1.CutCommand,
    'image': imageCommand_1.ImageCommand,
    'qr': qrCommand_1.QRCommand,
};
